engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/rollinggroup/RollingGroupOperator.java

```diff
@@ -0,0 +1,459 @@
+            @Nullable final String timestampColumnName,
+            final long reverseWindowScaleUnits,
+            final long forwardWindowScaleUnits,
+            final ColumnSource<?>[] valueSources
+    // region extra-constructor-args
+    // endregion extra-constructor-args
+    ) {
+        super(pairs[0], affectingColumns, rowRedirection, timestampColumnName, reverseWindowScaleUnits,
+                forwardWindowScaleUnits, true);
+
+        inputColumnNames = new String[pairs.length];
+        outputColumnNames = new String[pairs.length];
+        outputSources = new ColumnSource[pairs.length];
+        outputSourceMap = new HashMap<>();
+
+        // For the sake of rolling group operators, we need to map from every row to its bucket row set (or the
+        // non-null version) and to the start and end ranges of keys for the window slice. If we are using
+        // redirection, use it for these structures as well.
+        if (rowRedirection != null) {
+            innerGroupRowSetSource = new ObjectArraySource<>(RowSet.class);
+            groupRowSetSource = WritableRedirectedColumnSource.maybeRedirect(rowRedirection, innerGroupRowSetSource, 0);
+
+        } else {
+            groupRowSetSource = new ObjectSparseArraySource<>(RowSet.class);
+            innerGroupRowSetSource = null;
+        }
+
+        // If we are computing through ticks, we won't need sources for start and end.
+        if (timestampColumnName == null) {
+            innerStartSource = null;
+            innerEndSource = null;
+            startSource = null;
+            endSource = null;
+        } else {
+            // We are computing using the timestamp columns, create the appropriate sources (potentially redirected)
+            if (rowRedirection == null) {
+                startSource = new LongSparseArraySource();
+                endSource = new LongSparseArraySource();
+                innerStartSource = null;
+                innerEndSource = null;
+            } else {
+                innerStartSource = new LongArraySource();
+                innerEndSource = new LongArraySource();
+                startSource = WritableRedirectedColumnSource.maybeRedirect(rowRedirection, innerStartSource, 0);
+                endSource = WritableRedirectedColumnSource.maybeRedirect(rowRedirection, innerEndSource, 0);
+            }
+        }
+
+        for (int ii = 0; ii < pairs.length; ii++) {
+            final MatchPair pair = pairs[ii];
+            final Class<?> csType = valueSources[ii].getType();
+
+            inputColumnNames[ii] = pair.rightColumn;
+            outputColumnNames[ii] = pair.leftColumn;
+
+            if (csType == Boolean.class || csType == boolean.class) {
+                ColumnSource<Byte> reinterpreted =
+                        (ColumnSource<Byte>) ReinterpretUtils.maybeConvertToPrimitive(valueSources[ii]);
+
+                outputSources[ii] = timestampColumnName != null
```

Comment:
I would write two factory methods.
One for the source-driven version, one for the position-driven one.
Then, you can invoke the position-driven one once, and only have the funky (but correct) math once, and add a comment for it.
OK to add more maps to the `FactoryHelper` in `AggregateColumnSource`.